core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java

```diff
@@ -517,40 +540,54 @@ private HttpResponse newResponse(HttpMethod method, boolean isHealthy) {
         }
     }
 
-    private AggregatedHttpResponse getResponse(boolean isHealthy) {
-        if (isHealthy) {
-            return healthyResponse;
-        }
-
-        if (serverStopping) {
-            return stoppingResponse;
+    private AggregatedHttpResponse getResponse(HealthStatus healthStatus) {
+        switch (healthStatus) {
+            case HEALTHY:
+                return healthyResponse;
+            case DEGRADED:
+                return degradedResponse;
+            case STOPPING:
+                return stoppingResponse;
+            case UNDER_MAINTENANCE:
+                return underMaintenanceResponse;
+            default:
+                return unhealthyResponse;
         }
-
-        return unhealthyResponse;
     }
 
     private void onHealthCheckerUpdate(HealthChecker unused) {
         assert healthCheckerListener != null : "healthCheckerListener is null.";
-        assert pendingHealthyResponses != null : "pendingHealthyResponses is null.";
-        assert pendingUnhealthyResponses != null : "pendingUnhealthyResponses is null.";
+        assert pendingResponses != null : "pendingResponses is null.";
 
-        final boolean isHealthy = isHealthy();
+        final HealthStatus healthStatus = healthStatus();
         final PendingResponse[] pendingResponses;
         lock.lock();
         try {
-            final Set<PendingResponse> set = isHealthy ? pendingHealthyResponses
-                                                       : pendingUnhealthyResponses;
+            final List<PendingResponse> set =
+                    this.pendingResponses.stream()
+                                         .filter(res -> {
+                                             if (res.pollHealthStatus) {
+                                                 return res.interestedStatus != healthStatus;
```

Comment:
Shouldn't it filter when
- the `healthStatus` is changed from its previous `healthStatus`
- `res.interestedStatus == healthStatus`;